### PR TITLE
Shelley integration

### PR DESCRIPTION
**Pull request by nc6**

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+-- | TPraos parameters that are node independent
+data TPraosParams = TPraosParams {
+      -- | Active slots coefficient. This parameter represents the proportion of
+      -- slots in which blocks should be issued. This can be interpreted as the
+      -- probability that a party holding all the stake will be elected as
+      -- leader for a given slot.
+      tpraosLeaderF       :: Double
+    , tpraosSecurityParam :: SecurityParam
+    , tpraosEpochInfo     :: EpochInfo Identity
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs]: I'm commenting on this here, but maybe "here" is just one occurrence of many, and we need to fix this in a more fundamental way. This `EpochInfo Identity` is problematic. It means that we must know, _statically_, the length of every epoch (in slots). _We don't know that_. Even if every ledger (Shelley, Byron, ..) has a static number of slots per epoch (i.e., it's not an updatable protocol parameters, the exact _time_ of the hard fork _does_ depend on the ledger and so we cannot produce an `EpochInfo Identity`. This is wrecking my head a little however. I guess we can merge this as is, and re-think `EpochInfo` when it comes to the hard fork combinator.

**Comment by edsko** [comment]
One thought: perhaps we can implement something like `snapshot: EpochInfo IO -> EpochInfo Identity` which would produce an EpochInfo "for the epochs we currently know of" and throw an `error` for any epochs later than that. T

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Shelley/Ledger.hs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+  newtype LedgerConfig ShelleyBlock = ShelleyLedgerConfig SL.Globals
+    deriving (Generic)
+
+  applyChainTick
+    (ShelleyLedgerConfig globals)
+    slotNo
+    (ShelleyLedgerState pt history bhState) =
+      TickedLedgerState
+        . ShelleyLedgerState pt history
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Shelley/Ledger.hs]: Ok, I take this back. After https://github.com/input-output-hk/ouroboros-network/pull/1562 is merged this code is correct as-is.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Shelley/Block.hs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    ShelleyHeader
+      <$> fromCBOR
+      <*> fromCBOR
+
+encodeShelleyBlockWithInfo :: ShelleyBlock -> BinaryInfo Encoding
+encodeShelleyBlockWithInfo blk@(ShelleyBlock (Block h _)) = BinaryInfo
+  { binaryBlob = toCBOR blk
+  , headerOffset = 1 -- 'encodeListLen' of the block: header + body + wits
+    -- TODO When we have annotations in the Shelley decoders, we will not
+    -- need to do this
```
[the remark this answers is not in the document]
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Shelley/Block.hs]: Indeed

**Comment by mrBliss** [comment]
This will need rebasing after #1657 is merged. Everything from this PR should move to `ouroboros-consensus-shelley`, see `ouroboros-consensus-byron` for a template.

**Comment by mrBliss** [comment]
Superseded by #1695. I will bring over the tests to that PR too.

**Pull request closed without merging**